sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -548,6 +698,64 @@ graph LR
     Modifiable -->|Finalize| Executable
 ....
 
+==== Executable Graph Update
+
+A graph in the executable state can have the configuration of its nodes modified
+using a concept called graph _update_. This avoids a user having to rebuild and
+finalize a new executable graph when only the inputs & outputs to a graph
+change between submissions. Updates to a graph take effect from the next
+submission of that graph and will not affect any previous submissions or
+in-flight executions of the same graph.
+
+The only type of nodes that are currently supported for updating in a graph are
+kernel execution nodes.
+
+The aspects of a kernel execution node that can be configured during update are:
+
+* Parameters to the kernel.
+* Execution ND-Range of the kernel.
+
+To update an executable graph, the `property::graph::updatable` property must
+have been set when the graph was created during finalization. Otherwise, an
+exception will be thrown if a user tries to update an executable graph. This
+guarantee allows the backend to provide a more optimized implementation, if
+possible.
```

Comment:
@sommerlukas As part of this PR should we specify in `sycl_ext_oneapi_graph_fusion` that `require_fusion` and `updatable` can't both be passed on finalization?

Reply:
Depends a bit on how strict you want to be about this sentence for fused graphs:
>  This avoids a user having to rebuild and finalize a new executable graph when only the inputs & outputs to a graph change between submissions

In general, not every update necessarily requires recompilation of the fused kernel (we also have a caching mechanism for the existing kernel fusion extension).

The following changes to a graph would definitely require re-compilation of the fused kernel:
* Changes of the internalization configuration of a buffer/accessor/pointer during parameter update.
* Two parameters that had identical value have different values after parameter update.
* The value of a scalar parameter changes through parameter update.
* The size of local/private memory per work-item for internalization changes through an ND-range update.
* The ND-range update of two kernels leads to the kernel requiring some sort of remapping due to heterogeneous ND-ranges after the update.

I think there are two potential solutions for this:
* We disallow combination of `updatable` and `requires_fusion` (or fusion in general) altogether.
* We allow the combination, but `command_graph<...>::update` would throw an error if the updated graph would require re-compilation of the fused kernel. Potentially, we could re-use some of the existing caching logic to determine that.

What's your preference?

Reply:
> The ND-range update of two kernels leads to the kernel requiring some sort of remapping due to heterogeneous ND-ranges after the update.

This will only lead to a cache miss if not all the ND-ranges are equal. Otherwise, we do not need to recompile.

Reply:
> I think there are two potential solutions for this:
> 
>    * We disallow combination of `updatable` and `requires_fusion` (or fusion in general) altogether.
> 
>    * We allow the combination, but `command_graph<...>::update` would throw an error if the updated graph would require re-compilation of the fused kernel. Potentially, we could re-use some of the existing caching logic to determine that.
> 
> 
> What's your preference?

The second option seems better as a user can benefit to both optimizations, so I'd go for that at the moment while `sycl_ext_oneapi_graph_fusion` is still in the proposed status. But if later on as part of implementing graph fusion something unforeseen come up then we can fall back to option 1.

Reply:
Yes, I agree. Makes sense to leave the option open for now, we can always introduce that restriction later if implementation experience shows that the combination of update and fusion is not possible.